gtars/src/bbcache/client.rs

```diff
@@ -0,0 +1,206 @@
+use anyhow::{Context, Result};
+use std::collections::HashMap;
+use std::fs::{create_dir_all, read_dir, remove_dir, remove_file};
+use std::io::{Error, ErrorKind};
+use std::path::{Path, PathBuf};
+use walkdir::WalkDir;
+
+use log::info;
+
+use super::consts::{DEFAULT_BEDBASE_API, DEFAULT_BEDFILE_EXT, DEFAULT_BEDFILE_SUBFOLDER};
+use super::utils::{bb_url_for_regionset, get_abs_path};
+use crate::common::models::region_set::RegionSet;
+
+pub struct BBClient {
+    pub cache_folder: PathBuf,
+    pub bedbase_api: String,
+}
+
+impl BBClient {
+    pub fn new(cache_folder: Option<PathBuf>, bedbase_api: Option<String>) -> Result<Self> {
+        let cache_folder = get_abs_path(cache_folder, Some(true));
+        let bedbase_api = bedbase_api.unwrap_or_else(|| DEFAULT_BEDBASE_API.to_string());
+
+        Ok(BBClient {
+            cache_folder,
+            bedbase_api,
+        })
+    }
+
+    pub fn load_bed(&mut self, bed_id: &str) -> Result<RegionSet> {
+        let bedfile_path = self.bedfile_path(bed_id, Some(false));
+
+        if bedfile_path.exists() {
+            info!("Loading cached BED file from {:?}", bedfile_path.display());
+            return RegionSet::try_from(bedfile_path.as_path());
+        }
+
+        let regionset = self.download_bed_file_from_bb(bed_id)?;
+        info!("Downloaded BED file from BEDbase: {}", bed_id);
+        // write(&bedfile_path, bed_data)?;
+        regionset.to_bed_gz(bedfile_path.clone().as_path())?;
+
+        Ok(regionset)
+    }
+
+    pub fn add_local_bed_to_cache(
+        &mut self,
+        bedfile: PathBuf,
+        force: Option<bool>,
```

Comment:
Pythons "optional named params" dont translate well to rust. Im curious what people think I've gone back and forth. Having to always pass `None` here seems less than ideal -- but maybe if its not user-facing, its fine.

Another way I've solved this is just separate into two functions:
`add_local_bed_to_cache` and `add_local_bed_to_cache_forced`